#277 feature derive build rs

Crzyrndm opened the pull request:
closes #277 

This uses a build script to generate internal conditional compilation gates. This allows keeping the cargo features clean, while internally being able to use readable checks like
```Rust
#[cfg(any(
    condition = "family_L4x1",
    condition = "family_L4x2",
    condition = "family_L4x3",
))]
```
or
```Rust
#[cfg(condition = "peripheral_usart3")]
```

The build script is dependency free and simple (just names and a boolean condition) so there is basically no impact on compile time and very little to understand. Currently generates two sets of "conditions"
```
condition = "peripheral_<name>"
condition = "family_<name>"
```
peripheral presense mapping is derived from the [STM32 Cube repository](https://github.com/STMicroelectronics/STM32CubeL4). I chose to map all peripherals (even those present on every device) for maximum clarity. the peripheral name used is the lower case of the peripheral #define from that repository

suggestions on naming very welcome (I consider "condition" to be a placeholder)

korken89 commented:
Very nice improvement!
Would you see this as ready for merge? Seems to work fine for me.

Crzyrndm commented:
Feature wise, I believe so. The list of peripherals and which MCU they are present on is fairly complete (based on the C HAL) and I can't see any major downsides outside of this being a somewhat obscure capability of the build script (and the peripheral list being somewhat long ofc).

The style of the gating I am less sure on.
* is 'condition' a good prefix (should it even have a prefix)?
* should peripheral/family names be uppercase to match what is in the C-HAL / PAC?
* is `family_L41_42` a good naming scheme for those groupings?
* should finer details (e.g. what type of RTC / USB peripheral is present) get their own grouping? This group is probably going to grow over time. To me it would make sense to at least have a separated list for it, not sure what prefix to use for 'RTC is type 3' or 'USB is device only'

Crzyrndm commented:
Coming back to this after a few weeks away. Updated naming, cleaned up the build script with an eye toward future additions and generally just looking with fresh eyes

Each gate type now has its own unique gate type rather than all using `condition = "prefix_..."`
* #[cfg(family = "L43_44")] // group by the second digit
* #[cfg(family = "L4x2")] // group by the third digit
* #[cfg(has_peripheral = "adc1")] // peripheral presence
* #[cfg(has_peripheral_variant = "rtc_type3")] // peripheral subsets / variants
Naming is still a little shaky but it made no sense to standardise the left side of the gate
